kernel/src/table_changes/scan.rs

```diff
@@ -0,0 +1,203 @@
+    /// 4` to return a subset of the rows in the scan which satisfy the filter. If `predicate_opt`
+    /// is `None`, this is a no-op.
+    ///
+    /// NOTE: The filtering is best-effort and can produce false positives (rows that should should
+    /// have been filtered out but were kept).
+    pub fn with_predicate(mut self, predicate: impl Into<Option<ExpressionRef>>) -> Self {
+        self.predicate = predicate.into();
+        self
+    }
+
+    /// Build the [`TableChangesScan`].
+    ///
+    /// This does not scan the table at this point, but does do some work to ensure that the
+    /// provided schema make sense, and to prepare some metadata that the scan will need.  The
+    /// [`TableChangesScan`] type itself can be used to fetch the files and associated metadata required to
+    /// perform actual data reads.
+    pub fn build(self) -> DeltaResult<TableChangesScan> {
+        // if no schema is provided, use snapshot's entire schema (e.g. SELECT *)
+        let logical_schema = self
+            .schema
+            .unwrap_or(self.table_changes.schema.clone().into());
+        let mut have_partition_cols = false;
+        let mut read_fields = Vec::with_capacity(logical_schema.fields.len());
+
+        // Loop over all selected fields and note if they are columns that will be read from the
+        // parquet file ([`ColumnType::Selected`]) or if they are partition columns and will need to
+        // be filled in by evaluating an expression ([`ColumnType::Partition`])
+        println!("Logical schema: {:?}", logical_schema);
+        let column_types = logical_schema
+            .fields()
+            .enumerate()
+            .map(|(index, logical_field)| -> DeltaResult<_> {
+                if self
+                    .table_changes
+                    .partition_columns()
+                    .contains(logical_field.name())
+                {
+                    // Store the index into the schema for this field. When we turn it into an
+                    // expression in the inner loop, we will index into the schema and get the name and
+                    // data type, which we need to properly materialize the column.
+                    have_partition_cols = true;
+                    Ok(ColumnType::Partition(index))
+                } else if CDF_FIELDS
+                    .iter()
+                    .any(|field| field.name() == logical_field.name())
+                {
+                    // CDF Columns are generated, so they do not have a column mapping.
+                    Ok(ColumnType::Selected(logical_field.name().to_string()))
+                } else {
+                    // Add to read schema, store field so we can build a `Column` expression later
+                    // if needed (i.e. if we have partition columns)
+                    let physical_field =
+                        logical_field.make_physical(*self.table_changes.column_mapping_mode())?;
```

Comment:
FYI this changes a lot to support column mapping and/or nested columns: https://github.com/delta-io/delta-kernel-rs/pull/512

Reply:
Hmm how do you think we should handle this?

Reply:
This is an area where CDF is not different from a normal scan. We just need to leverage the same approach the other PR introduces. IMO we should not solve it here and just let the other PR pick it up.

Reply:
sweet, ty!